### PR TITLE
feat: detect mutation to values of reactive vars

### PR DESCRIPTION
**Pull request by maartenbreddels**
A common source of error is mutation of values in reactive vars. The reactive var cannot notice a change in its value (e.g. a list) if the list is mutated in place.

A user can be mislead that it is working correctly, when another reactive variable triggers a rerender, teaching bad habits.

Detecting mutations in Python without using proxies can be done by making a deepcopy of the object, and comparing the reference to the deepcopy.

This comes at the cost of performance and memory usage, therefore we should enabled this by default in development mode (tests run in this mode by default), so app builders are aware of mistakes while developing.
In production mode we can disable the mutation checks and behave as before.

TODO:
 - [ ] Trigger calls to check_mutations() from several places: after an event handler (requires a change in reacton) and after a component run.
 - [ ] We probably do not want two equals function in solara/reacton, reconcile this.
 - [ ] Give an option to opt out of the mutation check per reactive var (e.g. when equals fails), or for performance reasons.
 - [ ] support reactive.get(copy=True) to always get a copy, even when _CHECK_MUTATIONS is False
 - [ ] Do we need support reactive.get(reference=True) to always get a reference, or is the opt-out enough?

### CODE REVIEW

**Comment by Corvince** [comment]
I don't know if this is the right place to discuss this, but since this PR bundles several issues, I'll give it a try hear.

First of all I would question using mutable data is per-se a bad habit. I think there a valid use-cases and in generally it can be much more intuitive. Maybe you can write a bit about why mutable data is bad in the context of solara.

Second, I worked the past year with Vue.js a lot and from that direction I now find the name very confusing. In vuejs `reactive(foo)` somewhat does the opposite from solara. It wraps any object into a mutable proxy, but it does not work on primitive values. In solara it works with any variable, but its much more intuitive with simple types, since they are immutable and so you can't accidentally mutate the state (which this PR addresses). So where does the name "reactive" come from and why not  for example signals?

Lastly, I like the approach of [Solid.js](https://docs.solidjs.com/reference/basic-reactivity/create-signal#equals), where you can create a signal with the option `{equal: false}`, which just always updates when the `set` function is called. This would allow mutable data with a clear trigger for updates.

Hope my question and comments are helpful

**Comment by maartenbreddels** [comment]
> Lastly, I like the approach of [Solid.js](https://docs.solidjs.com/reference/basic-reactivity/create-signal#equals), where you can create a signal with the option `{equal: false}`

Very interesting idea, and would solve https://github.com/widgetti/solara/issues/245

I'll start writing about this topic (hopefully soon).

**Comment by maartenbreddels** [comment]
> Second, I worked the past year with Vue.js a lot and from that direction I now find the name very confusing.

Vue's refs and Solara reactive's are signals. I personally don't like the name signals at all (very technical name), but I think it would be less confusing in the end if Vue and Solara used signals as the name.

Vue's reactive (a proxy around a signal/ref), can be done in Python as well, and I'm mentally almost there... :)

I would say mutation is not necessarily bad, but it's difficult to reason about code if lists, etc., are mutated while parallel written into a database (in a different thread). The code might crash if another thread is iterating over a list while another thread is mutating it. You may not have stored in the database what you thought, etc.

To understand a larger codebase, data mutation makes it more difficult to reason about your code because you need to know who mutates where and when.

So basically, we say, don't mutate your state/data, mutate the signals/reactives, since those can also be 'observed' (i.e. we know we need to rerender, and we do not have the risk of our UI being out of sync with the state).

We plan to support a custom equals operator (which could always return False to support #245), but there is a small caviat there as well: If the same (mutated) objects/arguments are also passed to a child component, we assume that the component does not need a rerender. So, to support mutating objects we also need to support custom equals for components (which we plan to do).

**Comment by maartenbreddels** [comment]
Above TODO's are not an issue, since this is an opt-in feature now.

File: solara/toestand.py [hunk 1]
```diff
@@ -268,23 +275,93 @@ def set(self, value: S):
     def initial_value(self) -> S:
         pass
 
+    def _check_mutation(self):
+        pass
+
+
+def _find_outside_solara_frame() -> Optional[FrameType]:
+    # the module where the call stack origined from
+    current_frame = inspect.currentframe()
```
Comment (iisakkirotko) [inline solara/toestand.py:284]: In interest of performance, should we use `sys._getframe`? Recall the discussion in https://github.com/widgetti/solara/pull/517, which references the testing done in https://gist.github.com/JettJones/c236494013f22723c1822126df944b12.

File: solara/toestand.py [hunk 2]
```diff
@@ -268,23 +275,93 @@ def set(self, value: S):
...
+    return module_frame
+
 
 class KernelStoreValue(KernelStore[S]):
     default_value: S
+    _traceback: Optional[inspect.Traceback]
+    _default_value_copy: Optional[S]
 
-    def __init__(self, default_value: S, key=None):
+    def __init__(self, default_value: S, key=None, equals: Equals = equals, unwrap=lambda x: x):
```
Comment (iisakkirotko) [inline solara/toestand.py:303]: Also naming legibility here.

File: solara/website/pages/roadmap/roadmap.md [hunk 3]
```diff
@@ -11,6 +11,9 @@ Exciting news! We aim to release Solara 2.0 by the end of the year. For the 2.0
 
 - Elimination of common mistakes, such as detecting state mutations and avoiding misuse of hooks (e.g., using hooks in loops).
 
+    State mutation detection will be the default for Solara 2.0, but can be enabled in Solara 1.x by setting the environment variable `SOLARA_STORAGE_MUTATION_DETECTION=1`.
```
Comment (iisakkirotko) [inline solara/website/pages/roadmap/roadmap.md:14]:
```suggestion
    State mutation detection will be the default for Solara 2.0, but can be enabled in Solara >= 1.41.0 by setting the environment variable `SOLARA_STORAGE_MUTATION_DETECTION=1`.
```

Comment (iisakkirotko) [reply solara/website/pages/roadmap/roadmap.md:14]: Since we're determined to get this in for next release :)

File: solara/_stores.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    @property
+    def lock(self):
+        return self._storage.lock
+
+    def get(self) -> S:
+        print("check!")
+        self.check_mutations()
+        self._ensure_public_exists()
+        value = self._storage.get()
+        assert value.public is not None
```
[the remark this answers is not in the document]
Comment (maartenbreddels) [reply solara/_stores.py:34]: In the dataclass we document this:
```
# this is the value that is exposed in .get(), it is a deep copy of private
```

But since we do this lazily, it can be None (_ensure_public_exists should make sure it doesn't, but this is to also get the typing right, and to self document)

File: solara/toestand.py [hunk 5]
```diff
@@ -199,8 +206,8 @@ class KernelStore(ValueBase[S], ABC):
     _type_counter: Dict[Any, int] = defaultdict(int)
     scope_lock = threading.RLock()
 
-    def __init__(self, key=None):
-        super().__init__()
+    def __init__(self, key=None, equals: Equals = equals):
```
[the remark this answers is not in the document]
Comment (maartenbreddels) [reply solara/toestand.py:209]: Should we replace Equals by the type? `Callable[[Any, Any], bool]` ? That makes the docstrings easier to read indeed.

File: solara/toestand.py (hunk 1 above)
Comment (maartenbreddels) [reply solara/toestand.py:284]: Sound good!

File: solara/toestand.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (iisakkirotko) [reply solara/toestand.py:209]: What do you think about also renaming the default value to distinguish it, i.e.
```python
from reacton.utils import equals as default_equals
```

[the remark this answers is not in the document]
Comment (maartenbreddels) [reply solara/toestand.py:209]: I like the idea of a better name, but default does not add much information. Our equals does a bit 'extra'. What about `equals_extra`?

[the remark this answers is not in the document]
Comment (iisakkirotko) [reply solara/toestand.py:209]: Sounds good!
